UsageEnvironment/Log.cpp

```diff
@@ -77,20 +86,127 @@ FILE* _Log::AcquireFile() {
+      }
+    }
+  }
+  closedir(backupDir);
+#endif
+}
+
+void _Log::BackupLog() {
+#if defined(LOG_FILE_BACKUP_DIR) && defined(LOG_FILE_TMP_DIR)
+  RemoveOldBackup();  // Remove old backup log if it exists (flash memory should never be at risk of getting filled up)
+  fprintf(stdout, "Attempting to backup log.\n");
+
+  // Backup new log file.
+  char newPath[PATH_MAX];
+  FILE* backupLogFile;
+  FILE* tmpLogFile;
+  size_t logFileSize;
+  char* fileBuffer;
+
+  // Formulate new path
+  snprintf(newPath, sizeof(newPath), LOG_FILE_BACKUP_DIR "/live555-%lu.log", time(NULL));
+  fprintf(stdout, "renaming %s to %s\n", logFilePath, newPath);
+
+  // Open backup file and current logging file
+  backupLogFile = fopen(newPath, "w");
+  tmpLogFile = fopen(logFilePath, "r");
+
+  // Get file size and create buffer for copying
+  fseek(tmpLogFile, 0, SEEK_END);
+  logFileSize = ftell(tmpLogFile);
+  fileBuffer = (char *)malloc(logFileSize);
+  rewind(tmpLogFile);
+
+  if (fileBuffer == NULL) {
+    fprintf(stderr, "Couldn't create temp file read buffer.\n");
+    goto _clean_up_file_copy;
+  }
+
+  {
+    // Read temp log file contents
+    size_t bytesRead = fread(fileBuffer, sizeof(char), logFileSize, tmpLogFile);
+    if (bytesRead != logFileSize) {
+      perror("Read incorrect number of bytes from tmp log file.");
+      goto _clean_up_file_copy;
+    }
+    {
+      // Write temp log file contents to backup log file
+      size_t bytesWritten = fwrite(fileBuffer, sizeof(char), logFileSize, backupLogFile);
+      if (bytesWritten != logFileSize) {
+        perror("Wrote incorrect number of bytes to backup log file.");
+        goto _clean_up_file_copy;
+      }
+      {
+        // Give everyone rw permissions for this backup file
+        mode_t filePerms = S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP | S_IROTH | S_IWOTH;
+        if (chmod(newPath, filePerms) != 0) {
+          perror("Error setting file permissions");
+          goto _clean_up_file_copy;
+        }
+      }
+    }
+  }
+_clean_up_file_copy:
+  // Flush all streams
+  fflush(stderr);
+  fflush(stdout);
+  fflush(backupLogFile);
+
+  // Close all files / free memory
+  fclose(tmpLogFile);
+  fclose(backupLogFile);
+  free(fileBuffer);
+#endif
+}
+
 void _Log::ReleaseFile(FILE* logFileHandle) {
   if( logFileHandle != NULL ) {
     fflush(logFileHandle);
-
+    struct stat file_stat;
     if( strcmp(logFilePath, "") != 0 ) {
-      // The file is not stdout or stderr, so we can close it
+      // The file is not stdout or stderr, so we can back it up if needed
+      // then close it
+      if( stat(logFilePath, &file_stat) < 0 ) {
+        perror("Error getting live555 log file stat.");
+        goto _dealloc_log_file_struct;
+      }
+      if( file_stat.st_size >= MAX_LOG_FILE_SIZE ) {
+#ifdef LOG_FILE_TMP_DIR
+        if( remove(logFilePath) < 0 ) {
+          perror("Error removing live555 tmp log in " LOG_FILE_TMP_DIR);
+        }
+#endif
+      }
+_dealloc_log_file_struct:
```

Comment:
Similarly here, we'll be dealing with multiple files, so will review this function once updated.